python/pylibcudf/pylibcudf/contiguous_split.pyx

```diff
@@ -0,0 +1,168 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+
+from cython.operator cimport dereference
+from libc.stdint cimport uint8_t, uintptr_t
+from libcpp.memory cimport make_unique, unique_ptr
+from libcpp.utility cimport move
+from pylibcudf.libcudf.contiguous_split cimport (
+    pack as cpp_pack,
+    packed_columns,
+    unpack as cpp_unpack,
+)
+from pylibcudf.libcudf.table.table cimport table
+from pylibcudf.libcudf.table.table_view cimport table_view
+
+from .gpumemoryview cimport gpumemoryview
+from .table cimport Table
+from .utils cimport int_to_void_ptr
+
+from types import SimpleNamespace
+
+import numpy as np
+
+
+cdef class PackedColumns:
+    """Column data in a serialized format.
+
+    Contains data from an array of columns in two contiguous buffers:
+    one on host, which contains table metadata and one on device which
+    contains the table data.
+
+    For details, see :cpp:class:`cudf::packed_columns`.
+    """
+    def __init__(self):
+        raise ValueError(
+            "PackedColumns should not be constructed directly. "
+            "Use one of the factories."
+        )
+
+    @staticmethod
+    cdef PackedColumns from_libcudf(unique_ptr[packed_columns] data):
+        """Create a Python PackedColumns from a libcudf packed_columns."""
+        cdef PackedColumns out = PackedColumns.__new__(PackedColumns)
+        out.c_obj = move(data)
+        return out
+
+    @property
+    def metadata(self):
+        """memoryview of the metadata (host memory)"""
+        cdef size_t size = dereference(dereference(self.c_obj).metadata).size()
+        cdef uint8_t* data = dereference(dereference(self.c_obj).metadata).data()
+        if size == 0:
+            return memoryview(np.ndarray(shape=(0,), dtype="uint8"))
+        return memoryview(
+            np.asarray(
+                SimpleNamespace(
+                    owner = self,
+                    __array_interface__ = {
+                        'data': (<uintptr_t>data, False),
+                        'shape': (size,),
+                        'typestr': '|u1',
+                        'strides': None,
+                        'version': 3,
+                    }
+                )
+            )
+        )
```

Comment:
To avoid the numpy dependence, could `PackedColumns` just advertise the buffer protocol directly? Then you could say `memoryview(packed_columns)` and get the metadata. Or return an object that advertises the buffer protcol: https://cython.readthedocs.io/en/latest/src/userguide/buffer.html

In python 3.12 and later we could just do that via the python `__buffer__` attribute, but we can easily do it in cython.

Reply:
Implemented `HostBuffer` that owns the data and implement the buffer protocol.

And we now have `PackedColumns.release()` to get the buffers. This means that we can free the two buffers individually.

Reply:
:+1: thanks!